internal/buffer/cursor.go

```diff
@@ -431,6 +431,127 @@ func (c *Cursor) WordLeft() {
 	c.Right()
 }
 
+// SubWordRight moves the cursor one sub-word to the right
+func (c *Cursor) SubWordRight() {
+	if (util.IsWhitespace(c.RuneUnder(c.X))) {
+		for util.IsWhitespace(c.RuneUnder(c.X)) {
+			if c.X == util.CharacterCount(c.buf.LineBytes(c.Y)) {
+				c.Right()
+				return
+			}
+			c.Right()
+		}
+		return
```

Comment:
So if it is on a whitespace, it jumps to the beginning of the nearest word, not to the next subword of this word? Looks inconsistent with the behavior of `WordRight` and `WordLeft`.

Reply:
Yes, it is inconsistent. The problem is that there is no established practice
how it is being done the best way. Let me elaborate on this a little bit more.

_Sublime Text_ does it in a consistent way. `SubWordLeft()` and `SubWordRight()` do feel here as being a subset of `WordLeft()` and `WordRight()`. The sub-word functions produce basically the same jump locations as the word variants but also taking camel-case into account and therefore just adding some additional jumps to that.

_VSCode_ doesn't do it in a consistent way. `SubWordLeft()` and `SubWordRight()` do produce additional jump locations the normal `WordLeft()` and `WordRight()` functions will never produce.
Thats what my implementation tries to mimic right now, because as you mentioned, it has some benefits to the other approach.

Actually the only consistency is that the current implementation of `WordLeft()` and `WordRight()` do work the same for _VSCode_, _Sublime_ and _micro_. So, lets not change `WordLeft()` and `WordRight()`.

Another example for this can be found in the docs for the [zsh line editor](https://zsh.sourceforge.io/Doc/Release/Zsh-Line-Editor.html).
There are 3 different implementations for `forward-word` (`vi-forward-word`, `emacs-forward-word`) all slightly working differently.

Reply:
Well, we can propose an updated algorithm for `WordLeft()` and `WordRight()` ... but isn't that out of scope for this PR?
I mean the PR is suppose to implement the sub-word behavior, not to revamp word behavior.

And to be honest, I actually do not feel comfortable with changing one of the application's main navigation methods along the way. I understand why you propose it, new features also always open the chance for small refactorings and stuff. But changing `WordLeft()` and `WordRight()`, feels like there is much at stake.

Reply:
I suppose there was no good reason for the current behavior of `WordLeft()` and `WordRight()` for non-word non-whitespace chars, it was just the easiest to implement. So I would take the risk and go ahead changing it.

But yeah, there is no strong need to do it in this PR, we can do it separately.

Still, how about changing the behavior of your `SubWordLeft()` and `SubWordRight()` for whitespace characters to be in line with `WordLeft()` and `WordRight()`?

Reply:
I guess there are good reasons for both ways... and that there is no real right or wrong, but just user preference instead.

For me, it is actually working the way I prefer it. Since _micro_, _vscode_ and _sublime_ are using the same behavior when it comes to word jumps and at least _micro_ and _vscode_ are sharing the same sub word jump behavior.

So, I'd vote for leaving both implementations (word and sub-word) as they are right now. Because, that's what would work best for me and hopefully for others as well.
Furthermore here are some advantages of the _vscode_ sub-word implementation. And whether you (in general, not you in particular) like _vscode_ or not, it is a very popular editor and I guess therefore you can say, that the _vscode_ sub-word implementation is at least a valid choice for users.
Besides this I already mentioned my concerns about changing an established (and probably well accepted) word jump behavior.

We could think of doing it the _zsh_ way... Implementing different word and sub-word behaviors and let the user choose by custom keybindings... Like: 

Probably my config:
```json
{
    "CtrlLeft": "SubWordLeftVSCode",
    "CtrlRight": "SubWordRightVSCode",
    "AltCtrlLeft": "WordLeft",
    "AltCtrlRight": "WordRight"
}
```
Probably the config of someone preferring _emacs_:
```json
{
    "CtrlLeft": "SubWordLeftEmacs",
    "CtrlRight": "SubWordRightEmacs",
    "AltCtrlLeft": "WordLeftEmacs",
    "AltCtrlRight": "WordRightEmacs"
}
```

And if someone feels like he needs to implement `SubWordLeft*` / `SubWordRight*` (`*` = emacs or whatever) he is free to do so.

Reply:
> Since _micro_, _vscode_ and _sublime_ are using the same behavior when it comes to word jumps

Ah, ok. If they all do that, then it's probably intentional, for a reason, not just out of developer's laziness. So then it makes sense to keep it as is.

Wait... I'm lazy to install vscode or sublime to check that, but, is this vscode too? https://vscode.dev/ And at least in this web version, as I can see, the behavior of word jumps (Ctrl-Left and Ctrl-Right) is NOT the same as in micro. With `foo == bar`, it jumps over `==`, not exactly the same way as subword jumps but close to it, and not at all like micro (which doesn't jump over `==` at all, it just crawls over each `=` character.)

> and at least _micro_ and _vscode_ are sharing the same sub word jump behavior.

Ok, if you are so attached to keeping exactly the same subword behavior as in vscode (including all corner cases), let's keep it.

(You said "Furthermore here are some advantages of the _vscode_ sub-word implementation", but didn't say what advantages... But anyway, ok, let's keep it, I'm getting tired of discussing this endlessly. :) )

Reply:
> Ah, ok. If they all do that, then it's probably intentional, for a reason, not just out of developer's laziness. So then it makes sense to keep it as is.

Well, I got myself confused with my own customized configs.
You were right. Word jumps are not the same in _micro_, _VSCode_ and _Sublime_ with default settings.
Therefore it was probably not for a good reason, but indeed developer's laziness. Sorry for the confusion.

> When it comes to the behavior when the cursor is on a non-word non-whitespace character, on the other hand, I think we should do quite the opposite (as I already said below). The present behavior of WordLeft() and WordRight() in this case doesn't make much sense to me. E.g. with foo == bar, is it really useful that WordRight moves to the first =, then to the second =, then to the whitespace, and only then jumps the word?

> I've checked how this behaves in nano, emacs and gedit. Gedit jumps to the next whitespace, skipping all consecutive non-word non-whitespace characters, e.g. jumps over == in this case (similarly to your SubWordRight), and nano and emacs jump to the next word right away, skipping all non-word characters including whitespace. Both behaviors make more sense to me than the current behavior of WordRight.

If you are still interested in taking the risk of changing `WordLeft()` and `WordRight()`, I slightly modified those functions to be in line with `SubWordLeft()` and `SubWordRight()`.
They are now mimicking _VSCode_'s word jump behavior, which can be indeed tested with  [https://vscode.dev/](https://vscode.dev/).

Reply:
Yeah, it looks good to me now, thanks.

Could you clean up the commit history in your PR: rebase it and squash the commits, except the change of `WordLeft()` and `WordRight()` (since that is a separate change, so it'd better be in a separate commit)?

Maybe, to keep the commit history most clean, make it 3 commits: 1. first rename `IsNonAlphaNumeric()` to `IsNonWordChar()`. 2. then all the changes related to subwords. 3. then the change of `WordLeft()` and `WordRight()`?

Reply:
Okay, no problem. The 3 commits got way cleaner, good idea.

Anything else left to do?